Add guides for Tizen Web ML API (#1432)
Samsung/tizen-docs: docs/application/web/guides/machine-learning/pipeline.md

```

As these are [privacy-related privileges](../../tutorials/sec-privileges.md) the application has to [request proper permissions using the PPM API](../security/privacy-related-permissions.md) (in [mobile](../../api/latest/device_api/mobile/tizen/ppm.html) and [wearable](../../api/latest/device_api/wearable/tizen/ppm.html) applications).

## Create and dispose machine learning inference pipelines
NNStreamer is a plugin for [GStreamer](https://gstreamer.freedesktop.org), adding tensor data types and enabling use of machine learning inference models in pipelines.
You can use some of the standard GStreamer and NNStreamer elements (also called _nodes_ throughout this guide) in your applications.

NNStreamer pipelines are created from string descriptions, for example:
```javascript
'appsrc ! other/tensor,dimension=(string)2:1:1:1,type=(string)int8,framerate=(fraction)0/1 ! tensor_filter framework=custom-easy model=my-custom-filter ! tensor_sink'
```
Exclamation marks `!` separate pipeline nodes.
You can learn how to define pipeline descriptions from [GStreamer documentation](https://gstreamer.freedesktop.org/documentation).

To create a machine learning inference pipeline:

> 1. Describe it as a string:
>    ```javascript
>    var pipelineDescription = 'videotestsrc num-buffers=3 ' +
>                              '! video/x-raw,width=20,height=15,format=BGRA ' +
>                              '! tensor_converter ' +
>                              '! fakesink';
>    ```
> 
> 2. Call `tizen.ml.pipeline.createPipeline()`:
>    ```javascript
>    var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
>    ```

Reviewer: Global: 
Please give one line space after a sentence is followed by code Like below:
What is **it** please can you replace it with actual subject.
```suggestion
1. Describe it as a string:

   ```javascript
   var pipelineDescription = 'videotestsrc num-buffers=3 ' +
                             '! video/x-raw,width=20,height=15,format=BGRA ' +
                             '! tensor_converter ' +
                             '! fakesink';
   ```

2. Call `tizen.ml.pipeline.createPipeline()`:

   ```javascript
   var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
   ```

Use `gst-inspect-1.0` to learn more about the differences between these to elements.

To get pipeline output:
1. Create a pipeline with a sink node that can pass data to application:
   ```javascript
   var pipelineDescription = 'videotestsrc num-buffers=3 ' +
                             '! videoconvert ' +
                             '! videoscale ' +
                             '! video/x-raw,format=RGBx,width=16,height=16,framerate=10/1 ' +
                             '! tensor_converter ' +
                             '! tensor_sink name=sinkx';
   var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
   ```

2. Register a listener triggered by data incoming to `sinkx`:
   ```javascript
   function sinkListener(sinkName, data) {
      console.log('SinkListener for ' + sinkName + ' sink called. Data dimensions: ' +
              data.tensorsInfo.getDimensions(0));

      // Read and process data according to your needs
   }
   pipeline.registerSinkListener('sinkx', sinkListener);
   ```

After starting the pipeline, `sinkListener` will be called repeatedly.

> ## Change data flow within pipeline
> You may want to set the the source of the data to one of several pipeline branches.
> It can be done with `input-selector` node and `Switch` API.
> 
> We will use the pipeline from the figure below to show how to use `input-selector`, to choose between blue and green branch as the source for the white branch.
> 
> ![Pipeline with branches and a switch](./media/pipeline_with_switch.svg)
> 
> To choose a source branch with `input-selector`:
> 1. Create a pipeline:
>    ```javascript
>    var pipelineDescription = 'input-selector name=ins ! tensor_converter ! tensor_sink name=sinkx ' + // white branch
>                              'videotestsrc is-live=true ! videoconvert ! ins.sink_0 ' + // blue branch
>                              'videotestsrc num-buffers=3 is-live=true ! videoconvert ! ins.sink_1'; // green branch
>    var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
>    ```
> 
> 2. Get `insSwitch`:
>    ```javascript
>    var insSwitch = pipeline.getSwitch('ins');
>    ```
> 
> 3. You can get `Switch` pads:
>    ```javascript
>    console.log(insSwitch.getPadList()); // ["sink_0", "sink_1"]
>    ```
> 
> 4. Choose the pad to be used as a source for `input-selector` and the white pipeline branch:
>    ```javascript
>    insSwitch.select('sink_1'); // green branch used as the source
>    ```

Reviewer: ```suggestion
## Change data flow within pipeline

You may want to set the the source of the data to one of several pipeline branches.
It can be done with `input-selector` node and `Switch` API. Use the pipeline from the following figure to show how to use `input-selector` to choose between blue and green branch as the source for the white branch.

![Pipeline with branches and a switch](./media/pipeline_with_switch.svg)

To choose a source branch with `input-selector`:

1. Create a pipeline:
  
    ```javascript
   var pipelineDescription = 'input-selector name=ins ! tensor_converter ! tensor_sink name=sinkx ' + // white branch
                             'videotestsrc is-live=true ! videoconvert ! ins.sink_0 ' + // blue branch
                             'videotestsrc num-buffers=3 is-live=true ! videoconvert ! ins.sink_1'; // green branch
   var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
   ```
2. Get `insSwitch`:
  
   ```javascript
   var insSwitch = pipeline.getSwitch('ins');
   ```

3. Get `Switch` pads:
   
    ```javascript
       console.log(insSwitch.getPadList()); // ["sink_0", "sink_1"]
   ```
4. Choose the pad to be used as a source for `input-selector` and the white pipeline branch:
   
      ```javascript
      insSwitch.select('sink_1'); // green branch used as the source
      ```

In the example above, we used an `input-selector` to choose from alternative sources.
Similarily, we can use an `output-selector` to direct data to one of a set of sinks.

## Start and stop data flow to a pipeline branch
Use `valve` element with `Valve` API to start and stop data flow in one of a pipeline branches.

To control data flow in a single pipeline branch:
1. Create and start a pipeline with a `valve` element:
   ```javascript
   var pipelineDescription = 'videotestsrc is-live=true ' +
                             '! videoconvert ' +
                             '! videoscale ' +
                             '! video/x-raw,format=RGBx,width=16,height=16,framerate=10/1 ' +
                             '! tensor_converter ' +
                             '! valve name=valve1 ' +
                             '! fakesink';

   var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
   pipeline.start();
   ```

> 
> 2. Get `Valve` element from the pipeline:
>    ```javascript
>    var valve = pipeline.getValve('valve1');
>    ```
> 
> 3. You can check if the valve is opened or closed:
>    ```javascript
>    console.log(valve.isOpen); // true
>    ```
> 
> 4. Set the desired valve state:
>    ```javascript
>    valve.setOpen(false); // stop the data flow

Reviewer: ```suggestion
2. Get `Valve` element from the pipeline:
   ```javascript
   var valve = pipeline.getValve('valve1');
   ```

3. You can check if the valve is opened or closed:
   ```javascript
   console.log(valve.isOpen); // true
   ```

4. Set the desired valve state:
   ```javascript

   valve.setOpen(false); // stop the data flow

   ```

> 

Reviewer: Kindly remove the extra line.

## Write custom data filters

> The data flowing through the pipeline can be transformed in application, using  `CustomFilter` callbacks.

Reviewer: Global : 1 line space between heading & next text following it 
Like this 
## Write custom data filters

The data flowing through the pipeline can be transformed in application, using  `CustomFilter` callbacks.

   > [!NOTE]
   > Sending data between `CustomFilter` running in JS application and GStreamer pipeline has large overhead.
   > Using `CustomFilter` with large tensors may result in unsatisfactory performance.

To transform the data within a callback registered in JS application:
1. Define information about `CustomFilter`'s input and output tensors:
   ```javascript
   var inputInfo = new tizen.ml.TensorsInfo();
   inputInfo.addTensorInfo('3D', 'UINT8', [4, 20, 15, 1]);
   var outputInfo = new tizen.ml.TensorsInfo();
   outputInfo.addTensorInfo('flat', 'UINT8', [1200]);
   ```

2. Register a `CustomFilter` in the pipeline. Remember about returning a proper status code from the callback - in this case we return `0` to indicate success:
   ```javascript
   function flattenFilter(inputData, outputData) {
       var rawInputData = inputData.getTensorRawData(0);
       outputData.setTensorRawData(0, rawInputData.data);

       return 0;
   };

   tizen.ml.pipeline.registerCustomFilter('flattenFilter', flattenFilter, inputInfo, outputInfo);

>    ```

Reviewer: To transform the data within a callback registered in JS application:

1. Define information about `CustomFilter`'s input and output tensors:
   
   ```javascript
   var inputInfo = new tizen.ml.TensorsInfo();
   inputInfo.addTensorInfo('3D', 'UINT8', [4, 20, 15, 1]);
   var outputInfo = new tizen.ml.TensorsInfo();
   outputInfo.addTensorInfo('flat', 'UINT8', [1200]);
   ```
2. Register a `CustomFilter` in the pipeline. Remember about returning a proper status code from the callback. In this case, `0` is returned, that indicates success:
  
 ```javascript
   function flattenFilter(inputData, outputData) {
       var rawInputData = inputData.getTensorRawData(0);
       outputData.setTensorRawData(0, rawInputData.data);
       return 0;
   };
   tizen.ml.pipeline.registerCustomFilter('flattenFilter', flattenFilter, inputInfo, outputInfo);
   ```

   > [!Note]
   > For performance reasons `inputData` and `outputData` passed to `flattenFilter` behave differently than ordinary `TensorsData` objects.

>    > They cannot be disposed and are invalidated when the callback returns - you have to copy them manually to use outside the callback.

Reviewer: The sentences are not clear in note, please rephrase. 
> [!NOTE]
   > For performance reasons, `inputData` and `outputData` are passed to `flattenFilter`,  behave differently than ordinary `TensorsData` objects.
   > They cannot be disposed and are invalidated when the callback returns - you have to copy them manually to use outside the callback.

   > `inputData` is read-only and `outputData` is initialized with random values.

> 3. Create a pipeline with a `custom-easy-filter` element:
>    ```javascript
>    var pipelineDescription = 'videotestsrc num-buffers=3 ' +
>                              '! video/x-raw,width=20,height=15,format=BGRA ' +
>                              '! tensor_converter ' +
>                              '! tensor_filter framework=custom-easy model=flattenFilter ' +
>                              '! fakesink';
>    var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
>    ```
> 
> When you will start the pipeline, the `flattenFilter` will transform 3-dimensional tensors into 1-dimensional vectors.

Reviewer: ```suggestion
3. Create a pipeline with a `custom-easy-filter` element:
   
    ```javascript
       var pipelineDescription = 'videotestsrc num-buffers=3 ' +
                             '! video/x-raw,width=20,height=15,format=BGRA ' +
                             '! tensor_converter ' +
                             '! tensor_filter framework=custom-easy model=flattenFilter ' +
                             '! fakesink';
      var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
      ```
When you start the pipeline, the `flattenFilter` transforms 3-dimensional tensors into 1-dimensional vector.

> 

Reviewer: Remove extra line

   ```javascript
   var pipelineDescription = 'appsrc name=srcx  ' +
                          '! other/tensor,dimension=(string)224:224:3:1,type=(string)float ' +
                          '! tensor_filter framework=tflite model=' + absoluteModelPath + ' ' +
                          '! appsink name=sinkx';
   var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
   ```

3. Set up other pipeline elements. We will use a custom `Source` and `Sink`:
   ```javascript
   var source = pipeline.getSource('srcx');

   function sinkListener(sinkName, data) {
     var rawData = data.getTensorRawData(0);
     console.log(sinkName + ' received the data:');
     console.log(rawData);
   };

   pipeline.registerSinkListener('sinkx', sinkListener);
   ```

4. Run the pipeline. We will use random input:
   ```javascript
   pipeline.start();

   var inputTensorsData = inputTensorsInfo.getTensorsData(0);
   var randomInput = [];
   for (var i = 0; i < 224 * 224 * 3; i++) {
       randomInput.push(Math.random());
   };
   inputTensorsData.setTensorRawData(0, randomInput);
   source.inputData(inputTensorsData);

>    ```

Reviewer: ## Use saved models

Using machine learning models trained with popular frameworks like [TensorFlow](https://tensorow.org) is one of the main use cases for NNStreamer pipelines.

To read a model from file and use it in a pipeline:

1. You have to know the absolute path to the saved model file. It can be obtained from a path relative to a virtual root:

   ```javascript
   var modelPath = 'documents/mobilenet_v1_1.0_224_quant.tflite';
   var URI_PREFIX = 'file://';
   var absoluteModelPath = tizen.filesystem.toURI(modelPath).substr(URI_PREFIX.length);
   ```
2. Create a pipeline with a `tensor_filter` node with `model` property set to the location of a model.
   
    ```javascript
       var pipelineDescription = 'appsrc name=srcx  ' +
                          '! other/tensor,dimension=(string)224:224:3:1,type=(string)float ' +
                          '! tensor_filter framework=tflite model=' + absoluteModelPath + ' ' +
                          '! appsink name=sinkx';
       var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
       ```
3. Set up other pipeline elements. We will use a custom `Source` and `Sink`:

    ```javascript
       var source = pipeline.getSource('srcx');
       function sinkListener(sinkName, data) {
       var rawData = data.getTensorRawData(0);
       console.log(sinkName + ' received the data:');
       console.log(rawData);
   };
   pipeline.registerSinkListener('sinkx', sinkListener);
   ```

4. Run the pipeline using  a random input:
  
     ```javascript
       pipeline.start();
       var inputTensorsData = inputTensorsInfo.getTensorsData(0);
       var randomInput = [];
       for (var i = 0; i < 224 * 224 * 3; i++) {
       randomInput.push(Math.random());
   };
   inputTensorsData.setTensorRawData(0, randomInput);
   source.inputData(inputTensorsData);
   ```suggestion

## Related Information
* Dependencies
  - Tizen 6.5 and Higher for Mobile
  - Tizen 6.5 and Higher for Wearable

>   - Tizen 6.5 and Higher for TV

Reviewer: ```suggestion
## Related information

- Dependencies
  - Tizen 6.5 and Higher for Mobile
  - Tizen 6.5 and Higher for Wearable
  - Tizen 6.5 and Higher for TV

## Create and dispose machine learning inference pipelines

NNStreamer is a plugin for [GStreamer](https://gstreamer.freedesktop.org).
It adds new pipeline nodes and data types that allow to run on-device machine learning inference.
You can use some of the standard GStreamer and NNStreamer elements (also referred to as _nodes_ throughout this guide) in your applications.

- NNStreamer pipelines are created from string descriptions, for example:

   ```javascript
   'appsrc ! other/tensor,dimension=(string)2:1:1:1,type=(string)int8,framerate=(fraction)0/1 ! tensor_filter framework=custom-easy model=my-custom-filter ! tensor_sink'
   ```

- Exclamation marks `!` separate pipeline nodes.

You can learn how to define pipeline descriptions from [GStreamer documentation](https://gstreamer.freedesktop.org/documentation).

To create a machine learning inference pipeline:

> 1. Describe it as a string:

Reviewer 2: Done

To get pipeline output:

1. Create a pipeline with a sink node that can pass data to application:

   ```javascript
   var pipelineDescription = 'videotestsrc num-buffers=3 ' +
                             '! videoconvert ' +
                             '! videoscale ' +
                             '! video/x-raw,format=RGBx,width=16,height=16,framerate=10/1 ' +
                             '! tensor_converter ' +
                             '! tensor_sink name=sinkx';
   var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
   ```

2. Register a listener triggered by data incoming to `sinkx`:

   ```javascript
   function sinkListener(sinkName, data) {
      console.log('SinkListener for ' + sinkName + ' sink called. Data dimensions: ' +
              data.tensorsInfo.getDimensions(0));

      // Read and process data according to your needs
   }
   pipeline.registerSinkListener('sinkx', sinkListener);
   ```

After starting the pipeline, `sinkListener` will be called repeatedly.

## Change data flow within pipeline

You may want to set the the source of the data to one of several pipeline branches.
It can be done with `input-selector` node and `Switch` API.

> Use the pipeline from the figure below to see how to use `input-selector` to choose between blue and green branch as the source for the white branch.

Reviewer: ```suggestion
You can get analyze the pipeline from the following figure to see how to use `input-selector` to choose between blue and green branch as the source for the white branch.

![Pipeline with branches and a switch](./media/pipeline_with_switch.svg)

To choose a source branch with `input-selector`:

1. Create a pipeline:

   ```javascript
   var pipelineDescription = 'input-selector name=ins ! tensor_converter ! tensor_sink name=sinkx ' + // white branch
                             'videotestsrc is-live=true ! videoconvert ! ins.sink_0 ' + // blue branch
                             'videotestsrc num-buffers=3 is-live=true ! videoconvert ! ins.sink_1'; // green branch
   var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
   ```

2. Get `insSwitch`:

   ```javascript
   var insSwitch = pipeline.getSwitch('ins');
   ```

3. You can get `Switch` pads:

   ```javascript
   console.log(insSwitch.getPadList()); // ["sink_0", "sink_1"]
   ```

4. Choose the pad to be used as a source for `input-selector` and the white pipeline branch:

   ```javascript
   insSwitch.select('sink_1'); // green branch used as the source
   ```

In the example above, we used an `input-selector` to choose from alternative sources.

> Similarily, we can use an `output-selector` to direct data to one of a set of sinks.

Reviewer: ```suggestion
In the example already discussed, we used an `input-selector` to choose from alternative sources.
Similarly, we can use an `output-selector` to direct data to one of a set of sinks.

   var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
   pipeline.start();
   ```

2. Get `Valve` element from the pipeline:

   ```javascript
   var valve = pipeline.getValve('valve1');
   ```

3. You can check if the valve is opened or closed:

   ```javascript
   console.log(valve.isOpen); // true
   ```

4. Set the desired valve state:

   ```javascript
   valve.setOpen(false); // stop the data flow
   ```

## Write custom data filters

The data flowing through the pipeline can be transformed in application, using  `CustomFilter` callbacks.

   > [!NOTE]
   > Sending data between `CustomFilter` running in JS application and GStreamer pipeline has large overhead.
   > Using `CustomFilter` with large tensors may result in unsatisfactory performance.

To transform the data within a callback registered in JS application:

1. Define information about `CustomFilter`'s input and output tensors:

   ```javascript
   var inputInfo = new tizen.ml.TensorsInfo();
   inputInfo.addTensorInfo('3D', 'UINT8', [4, 20, 15, 1]);
   var outputInfo = new tizen.ml.TensorsInfo();
   outputInfo.addTensorInfo('flat', 'UINT8', [1200]);
   ```

> 2. Register a `CustomFilter` in the pipeline. Remember about returning a proper status code from the callback. In this case, `0` is returned, that indicates success:

Reviewer: ```suggestion
2. Register `CustomFilter` in the pipeline. Ensure to return a proper status code from the callback. In this case, `0` is returned, that indicates success:

   ```javascript
   function flattenFilter(inputData, outputData) {
       var rawInputData = inputData.getTensorRawData(0);
       outputData.setTensorRawData(0, rawInputData.data);

       return 0;
   };
   tizen.ml.pipeline.registerCustomFilter('flattenFilter', flattenFilter, inputInfo, outputInfo);
   ```

   > [!Note]
   > `inputData` and `outputData` passed to `flattenFilter` are different than other `TensorsData` objects.
   > They cannot be disposed and are only valid within the callback - you have to copy them manually to use outside the callback.

>    > `inputData` is read-only and `outputData` is initialized with random values.

Reviewer: ```suggestion

   > [!Note]
   > `inputData` and `outputData` passed to `flattenFilter` are different than other `TensorsData` objects.
   > They cannot be disposed and are only valid within the callback. You have to copy them manually to use outside the callback.
   > `inputData` is read-only and `outputData` is initialized with random values.

3. Create a pipeline with a `custom-easy-filter` element:
   
   ```javascript
       var pipelineDescription = 'videotestsrc num-buffers=3 ' +
                             '! video/x-raw,width=20,height=15,format=BGRA ' +
                             '! tensor_converter ' +
                             '! tensor_filter framework=custom-easy model=flattenFilter ' +
                             '! fakesink';
      var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
   ```
When you start the pipeline, the `flattenFilter` transforms 3-dimensional tensors into 1-dimensional vector.

## Use saved models

Using machine learning models trained with popular frameworks like [TensorFlow](https://tensorow.org) is one of the main use cases for NNStreamer pipelines.

To read a model from file and use it in a pipeline:

> 1. You have to know the absolute path to the saved model file. It can be obtained from a path relative to a virtual root:

Reviewer: ```suggestion
1. You must know the absolute path to the saved model file. It can be obtained from a path relative to a virtual root:

   ```javascript
   var modelPath = 'documents/mobilenet_v1_1.0_224_quant.tflite';
   var URI_PREFIX = 'file://';

   var absoluteModelPath = tizen.filesystem.toURI(modelPath).substr(URI_PREFIX.length);
   ```

2. Create a pipeline with a `tensor_filter` node with `model` property set to the location of a model:

   ```javascript
   var pipelineDescription = 'appsrc name=srcx  ' +
                          '! other/tensor,dimension=(string)224:224:3:1,type=(string)float ' +
                          '! tensor_filter framework=tflite model=' + absoluteModelPath + ' ' +
                          '! appsink name=sinkx';
   var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
   ```

> 3. Set up other pipeline elements. We will use a custom `Source` and `Sink`:

Reviewer: ```suggestion
3. Set up other pipeline elements, and use custom `Source` and `Sink`:
